cli/provision.go

```diff
@@ -0,0 +1,261 @@
+//
+// Copyright (c) 2018
+// Mainflux
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package cli
+
+import (
+	"bufio"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/docker/docker/pkg/namesgenerator"
+	mfxsdk "github.com/mainflux/mainflux/sdk/go"
+	"github.com/spf13/cobra"
+)
+
+func printMap(title string, m map[string]string) {
+	fmt.Printf("\n%s\n", title)
+	for k, v := range m {
+		fmt.Printf("name: %s, id: %s\n", k, v)
+	}
+}
+
+type thing struct {
+	name  string
+	kind  string
+	id    string
+	token string
+}
+
+type channel struct {
+	name string
+	id   string
+}
+
+func createThing(name, kind, token string) (thing, error) {
+	loc, err := sdk.CreateThing(mfxsdk.Thing{Name: name, Type: kind}, token)
+	if err != nil {
+		return thing{}, err
+	}
+
+	// Received location header is in the format: /things/<thing_id>
+	id := strings.Split(loc, "/")[2]
+	t, err := sdk.Thing(id, token)
+	if err != nil {
+		return thing{}, err
+	}
+
+	m := thing{
+		name:  name,
+		kind:  kind,
+		id:    id,
+		token: t.Key,
+	}
+
+	return m, nil
+}
+
+func createChannel(name, token string) (channel, error) {
+	loc, err := sdk.CreateChannel(mfxsdk.Channel{Name: name}, token)
+	if err != nil {
+		return channel{}, nil
+	}
+
+	// Received location header is in the format: /channels/<channel_id>
+	id := strings.Split(loc, "/")[2]
+	c := channel{
+		name: name,
+		id:   id,
+	}
+
+	return c, nil
+}
+
+var cmdProvision = []cobra.Command{
+	cobra.Command{
+		Use:   "things",
+		Short: "things <things_csv> <user_token>",
+		Long:  `Provisions things`,
+		Run: func(cmd *cobra.Command, args []string) {
+			things := []thing{}
+
+			if len(args) != 2 {
+				logUsage(cmd.Short)
+				return
+			}
+
+			c, err := os.Open(args[0])
+			if err != nil {
+				logError(err)
+				return
+			}
+			reader := csv.NewReader(bufio.NewReader(c))
+			for {
+				l, err := reader.Read()
+				if err == io.EOF {
+					break
+				} else if err != nil {
+					logError(err)
+					return
+				}
+
+				m, err := createThing(l[0], l[1], args[2])
+				if err != nil {
+					logError(err)
+					return
+				}
+
+				things = append(things, m)
+			}
+
+			flush(things)
+		},
+	},
+	cobra.Command{
+		Use:   "channels",
+		Short: "channels <channels_csv> <user_token>",
+		Long:  `Provisions channels`,
+		Run: func(cmd *cobra.Command, args []string) {
+			channels := []channel{}
+
+			if len(args) != 2 {
+				logUsage(cmd.Short)
+				return
+			}
+
+			c, err := os.Open(args[0])
+			if err != nil {
+				logError(err)
+				return
+			}
+			reader := csv.NewReader(bufio.NewReader(c))
+			for {
+				l, err := reader.Read()
+				if err == io.EOF {
```

Comment:
You can use switch over error here, the code would be little more readable.

Reply:
Done. Looks readable.